Add support for Xilinx CANFD

smuthoju opened the pull request:
This patch set does the following:                                         
- Add Xilinx CANFD driver with CANFD support.                              
- Add device tree bindings and nodes for Xilinx CANFD controller.          
- Enable CANFD support for versalnet_rpu board configuration.              
- Add CONFIG_XILINX_CANFD guards to exclude unsupported CAN and CANFD tests.
- Update CAN counter sample with Xilinx CANFD specific handling.

Commit:
dts: bindings: can: Add Xilinx CANFD Controller

The Xilinx CANFD enhances traditional CAN communication by supporting
higher data rates and improved efficiency, making it suitable for
applications that require real-time data transfer, such as automotive
and industrial systems.

Xilinx CANFD supports data rates of up to 8 Mbps, allowing for faster
communication compared to standard CAN. The protocol supports longer
data frames, enabling the transmission of larger data packets and
compatible with both classical CAN and CANFD protocols.

Signed-off-by: Sai Priya M <saipriya.muthoju@amd.com>

Commit:
drivers: can: Add Xilinx CANFD driver support

Add support for Xilinx CANFD with CAN and CANFD functionality.

Features:
-Classic CAN 2.0A/B support (11-bit and 29-bit IDs).
-CANFD support with up to 64-byte data frames.
-Multiple operating modes: Normal, Loopback, Listen-only, Sleep.
-Configurable bit timing for arbitration and data phases.
-Single TX mailbox with interrupt-driven transmission.
-RX FIFO with multi-frame buffering.
-Comprehensive error handling and state management.
-Thread-safe operation with mutex protection.
-Power management with sleep/wake-up support.

Signed-off-by: Sai Priya M <saipriya.muthoju@amd.com>

Commit:
boards: amd: versalnet_rpu: Enable CANFD node

Add support for CANFD in Versal-NET RPU platform.

Signed-off-by: Sai Priya M <saipriya.muthoju@amd.com>

alexanderwachter reviewed (changes requested):
I have not read the entire code so far... just too many issues.

smuthoju commented on drivers/can/can_xilinx_canfd.c:
```diff
@@ -0,0 +1,1127 @@
+
+		if (len > 0 && len <= sizeof(frame.data)) {
+			for (int i = 0; i < len; i += 4) {
+				dw_offset = XCANFD_FRAME_DW_OFFSET(offset) +
```
You're right - recalculating the offset each iteration is unnecessary. I'll change it to increment the DW offset directly in the loop for better performance and clarity

smuthoju commented on drivers/can/can_xilinx_canfd.c:
```diff
@@ -0,0 +1,1127 @@
+	struct xcanfd_data *data = dev->data;
+	uint32_t isr;
+
+	if (dev == NULL) {
```
Yes its not required . i will remove it.

smuthoju commented on drivers/can/can_xilinx_canfd.c:
```diff
@@ -0,0 +1,1127 @@
+
+	if (isr & XCANFD_IXR_ARBLST_MASK) {
+		xcanfd_write32(config, XCANFD_ICR_OFFSET, XCANFD_IXR_ARBLST_MASK);
+		k_mutex_lock(&data->inst_mutex, K_FOREVER);
```
okay - using k_mutex_lock() in ISR context is not allowed. I'll remove the mutex usage from the ISR.

Commit:
dts: amd: versalnet: Add device tree support for Xilinx CANFD

Add CANFD nodes to device tree for Versal-NET platform.

Signed-off-by: Sai Priya M <saipriya.muthoju@amd.com>

Commit:
soc: xlnx: versalnet: Enable clock control for CAN support

Select CLOCK_CONTROL and CLOCK_CONTROL_FIXED_RATE_CLOCK to meet
the Xilinx CANFD driver dependencies for proper clock management.

Signed-off-by: Sai Priya M <saipriya.muthoju@amd.com>
